service/src/java/org/apache/hive/service/server/HiveServer2.java

```diff
@@ -895,6 +897,46 @@ private void stopOrDisconnectTezSessions() {
     }
   }
 
+  public synchronized void decommission() {
+    LOG.info("Decommissioning HiveServer2");
+    // Remove this server instance from ZooKeeper if dynamic service discovery is set
+    if (serviceDiscovery && !activePassiveHA && zooKeeperHelper != null) {
+      try {
+        zooKeeperHelper.removeServerInstanceFromZooKeeper();
+      } catch (Exception e) {
+        LOG.error("Error removing znode for this HiveServer2 instance from ZooKeeper.", e);
+      }
+    }
+    super.decommission();
+  }
+
+  public synchronized void graceful_stop() {
+    try {
+      decommission();
+      long maxTimeForWait = HiveConf.getTimeVar(getHiveConf(),
+              HiveConf.ConfVars.HIVE_SERVER2_GRACEFUL_STOP_TIMEOUT, TimeUnit.MILLISECONDS);
+      if (maxTimeForWait > 0) {
+        ExecutorService service = Executors.newSingleThreadExecutor();
+        Future future = service.submit(() -> {
+          while (getCliService() != null && getCliService().getSessionManager()
```

Comment:
Thank you for the review! wait/notify adds extra synchronizing codes in OperationManager. For this case, how about sleeping some time while looping? only a little time, say at most 100ms is wasted for waiting for OperationManager to be done, this does not bring much overhead to graceful stop.

Reply:
Agreeing that wait/notify may not be required in this case since this code path will only be executed when HS2 is being terminated. I am fine with using sleep. It would be good to add a comment in the code mentioning the same.

Reply:
Done, thank you!

Reply:
done, thank you!

Reply:
for timeout purpose, we can seperate the timeout checking from the waiting